src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -135,40 +152,92 @@ public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams reques
                 Diagnostics = mappedDiagnostics,
                 HostDocumentVersion = documentVersion,
             };
+        }
+
+        private async Task<Diagnostic[]> FilterHTMLDiagnosticsAsync(
+            Diagnostic[] unmappedDiagnostics,
+            RazorCodeDocument codeDocument,
+            DocumentSnapshot documentSnapshot)
+        {
+            var sourceText = await documentSnapshot.GetTextAsync();
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var classifiedSpans = syntaxTree.GetClassifiedSpans();
 
-            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
-            static bool CanDiagnosticBeFiltered(RazorLanguageKind kind, Diagnostic d)
+            var processedAttributes = new Dictionary<TextSpan, bool>();
+
+            var filteredDiagnostics = unmappedDiagnostics
+                .Where(d =>
+                    // TODO: undesired HTML Diagnostic codes
+                    // (https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401) ||
+                    !InAttributeContainingCSharp(d, sourceText, syntaxTree))
+                .ToArray();
+
+            return filteredDiagnostics;
+
+            bool InAttributeContainingCSharp(Diagnostic d, SourceText sourceText, RazorSyntaxTree syntaxTree)
             {
-                return kind == RazorLanguageKind.CSharp ?
-                    CanCSharpDiagnosticBeFiltered(d) :
-                    CanHTMLDiagnosticBeFiltered(d);
+                // Examine the _end_ of the diagnostic to see if we're at the
+                // start of an (im/ex)plicit expression. Looking at the start
+                // of the diagnostic isn't sufficient.
+                var absoluteIndex = d.Range.End.GetAbsoluteIndex(sourceText);
+                var change = new SourceChange(absoluteIndex, 0, string.Empty);
+                var owner = syntaxTree.Root.LocateOwner(change);
+
+                while (owner != null)
+                {
+                    if (owner is MarkupAttributeBlockSyntax markupAttributeNode)
+                    {
+                        if (processedAttributes.TryGetValue(markupAttributeNode.FullSpan, out var doesAttributeContainsCSharp))
+                        {
+                            return doesAttributeContainsCSharp;
+                        }
+
+                        doesAttributeContainsCSharp = CheckIfAttributeContainsCSharp(markupAttributeNode.FullSpan, classifiedSpans);
```

Comment:
Hmm I wouldn't use the classified spans to know if there's C# in the attribute. I'd rely on the syntax tree's representation. Aka if there's anything other then a normal attribute value syntax node then something funky is going on IIRC.

Reply:
Hmm, so I had that in the first implementation of this: https://github.com/dotnet/aspnetcore-tooling/pull/2973/commits/62c3c60b52f97a05227b244518c3b430187035e0#diff-4233fb067b526a5ca9d3d24db6fe013e92c3d7bb15b60b597503fb7d3a4eca4eR194-R215

Essentially the way that worked is for each diagnostic we go up the syntax tree till we hit the markup attribute node. If we hit CSharp before we hit the markup attribute node, we ignore that diagnostic. We also cache that markup attribute node, so that should it be hit again, we can automatically exclude that diagnostic as well. Note this meant diagnostics are hidden for the rest of the attribute value after the first implicit/explicit expression. 

I changed it to use the classified spans method so we don't require the diagnostic to be next to the implicit/explicit expression to be able to disable diagnostics. By using classified spans, we also ensure we don't get diagnostics for half the attribute value (before the implicit/explicit expression) and not for the rest, which could be confusing.

Let me know if you'd like me to revert https://github.com/dotnet/aspnetcore-tooling/pull/2973/commits/13437789c2bf26cff362f111102083d3bcc9d269.

Reply:
Discussed offline; updated https://github.com/dotnet/aspnetcore-tooling/pull/2973/commits/14f6f15361546a0311695dc93157842aa809dd69

![image](https://user-images.githubusercontent.com/14852843/105253386-5ca72780-5b4d-11eb-984f-c65fc184e5dc.png)